feat(lsp): inlay hints

dchinmay2 opened the pull request:
Implements inlay hints by adding automatic refresh and a public interface on top of #23736

Closes #18086

dchinmay2 committed:
feat(lsp): inlay hints

This PR adds automatic refresh and a public interface on top of #23736

Closes #18086

glepnir commented on runtime/lua/vim/lsp/_inlay_hint.lua:
```diff
@@ -149,26 +149,78 @@ end
+        if not M.__buffers[cb_bufnr].enabled then
+          return true
+        end
+        reset_timer(cb_bufnr)
```
not sure is the good way create timer for each buffer ,  and in `on_lines`  callback has `changedtick` param you can use  pass to request like `make_request(cb_bufnr, changedtick)`,  and when  lsp request callback running check `nvim_get_buf_chagnetick` make a compare. `if before_changetick < nvim_get_buf_changetick` do defer reqeust again.   some undo/redo also increase `changetick` and also some change without text chang also increase `changtick` .  this is my thought  and I do it on documentsymbol request . it combine 1000ms change to one request . 
 https://github.com/nvimdev/lspsaga.nvim/blob/3546dca5560b6868612acc567f6f3c07cf61c076/lua/lspsaga/symbol/init.lua#L25

p00f replied:
I don't like this either, is there a better way of remembering which buffers have inlay hint enabled (to handle refresh request)?

p00f replied:
That is what I was doing, except how do you cancel pending requests?

glepnir replied:
i had point it out `changetick ` increase without text changed and this change will not trigger `on_lines` callback . and there has `on_changedtick` in `nvim_buf_attach` . these problem  i had comment in prev :-)

dchinmay2 committed:
fix tests

jdrouhard commented on runtime/lua/vim/lsp/_inlay_hint.lua:
```diff
@@ -149,26 +149,78 @@ end
+  if not (M.__buffers[bufnr] and M.__buffers[bufnr].enabled) then
+    M.__buffers[bufnr] = { enabled = true, timer = nil }
+    M.refresh({ bufnr = bufnr })
+    api.nvim_buf_attach(bufnr, true, {
```
You'll also want handlers for `on_reload` and `on_detach`. `on_reload` should clear the cache and namespace, and refire a request. `on_detach` should clear out any cache related to that buffer or there will be a memory leak as more buffers accumulate state.

You may also want an `LspDetach` autocmd that does the same thing (in case LSP is detached from a buffer manually).

p00f replied:
done

jdrouhard commented on runtime/lua/vim/lsp/_inlay_hint.lua:
```diff
@@ -17,11 +17,10 @@ local hint_cache_by_buf = setmetatable({}, {
 local namespace = api.nvim_create_namespace('vim_lsp_inlayhint')
 
-M.__explicit_buffers = {}
+M.__buffers = {}
```
is this used outside this module any more?

p00f replied:
done

dchinmay2 committed:
address comments from PR review

 - add on_reload, on_detach handlers in `enable()` buf_attach, and
LspDetach autocommand in case of manual detach
 - change M.__buffers to __buffers

p00f commented on runtime/lua/vim/lsp/_inlay_hint.lua:
```diff
@@ -149,26 +149,78 @@ end
-    api.nvim_buf_clear_namespace(iter_bufnr, namespace, 0, -1)
   end
+  api.nvim_buf_clear_namespace(bufnr, namespace, 0, -1)
   vim.cmd('redraw!')
```
done

p00f commented on runtime/lua/vim/lsp/_inlay_hint.lua:
```diff
@@ -149,26 +149,78 @@ end
+function M.disable(bufnr)
+  bufnr = resolve_bufnr(bufnr or 0)
+  M.__buffers[bufnr].enabled = false
+  M.__buffers[bufnr].timer = nil
```
Will do

p00f replied:
done

lewis6991 commented on runtime/doc/lsp.txt:
```diff
@@ -699,6 +699,24 @@ buf_detach_client({bufnr}, {client_id})          *vim.lsp.buf_detach_client()*
+
+    Parameters: ~
+      • {bufnr}  (integer|nil) Buffer handle, or nil for current
+
```
Do we really need 4 functions for this?

p00f replied:
Do you want only a toggle function?

p00f replied:
Aren't enable and disable necessary for scripting

p00f replied:
And toggle is necessary if you want a mapping to toggle

p00f replied:
done

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -699,6 +699,24 @@ buf_detach_client({bufnr}, {client_id})          *vim.lsp.buf_detach_client()*
       • {bufnr}      (integer) Buffer handle, or 0 for current
       • {client_id}  (integer) Client id
 
+buf_inlay_hint_disable({bufnr})             *vim.lsp.buf_inlay_hint_disable()*
```
Why doesn't this live in `vim.lsp.buf.*` like almost all other buffer-relative LSP features? Strange that it has a `buf_` prefix instead.

p00f replied:
I'll move it; I was suggested `buf_`

justinmk replied:
still wondering where the `buf_` suggestion came from..

p00f replied:
It came from clason

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -699,6 +699,24 @@ buf_detach_client({bufnr}, {client_id})          *vim.lsp.buf_detach_client()*
+    Parameters: ~
+      • {bufnr}  (integer|nil) Buffer handle, or nil for current
+
+buf_inlay_hint_enable({bufnr})               *vim.lsp.buf_inlay_hint_enable()*
```
- "enable" can take an `enable` parameter, then you don't need the `_disable` variant.
- won't be long before someone asks for a getter (`is_enabled`), right? So if we add that now, then we also don't need `toggle()`.

p00f replied:
What if we don't add a getter and make it so that it toggles by default if there is no `enable` parameter, and enables/disables if there is?

justinmk replied:
Anything that results in 1 function instead of 3 is good enough.

p00f replied:
done

justinmk commented:
When I suggested skipping the public API in https://github.com/neovim/neovim/pull/23736 , my thinking was: just enable it by default. It doesn't need to be configurable.

Does LSP actually require the enable/disable stuff to be configured client-side? Seems like this could/should be a setting that could be sent to the server, using a standard "configure" LSP request. Does LSP have something like that?

justinmk commented on runtime/doc/news.txt:
```diff
@@ -84,8 +84,9 @@ The following new APIs or features were added.
 
-• Added |lsp-handler| for inlay hints: `textDocument/inlayHint` and
-  `workspace/inlayHint/refresh`
+• Implemented LSP inlay hints - |vim.lsp.buf_inlay_hint_enable()|,
```
news.txt (at minimum) is a great place to link to reference URLs such as https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_inlayHint

p00f replied:
done

clason commented:
> When I suggested skipping the public API in #23736 , my thinking was: just enable it by default. It doesn't need to be configurable.
> 
> Does LSP actually require the enable/disable stuff to be configured client-side? Seems like this could/should be a setting that could be sent to the server, using a standard "configure" LSP request. Does LSP have something like that? Can the client-side "capabilities" map be used to enable/disable inlay hints?

Inlay hints can be quite obtrusive (even moreso than diagnostics); they should definitely be under the user's control. (I refer to the semantic tokens decision, which triggered a deluge of "my highlight is suddenly broken!!eleven!" issues for nvim-treesitter.)

justinmk commented:
> Inlay hints can be quite obtrusive (even moreso than diagnostics); they should definitely be under the user's control. (I refer to the semantic tokens decision,

- Does the "client capabilities" give control or not?
- I don't see any other `enable()` methods in `:help lsp`. Why is this the only one?

clason commented:
> Does the "client capabilities" give control or not?

Not enough; that is all-or-nothing. The ask here is to only view the inlay hints on demand. not disable them completely

> I don't see any other enable() methods in :help lsp. Why is this the only one?

Because we

1.  started with the simple, non-controversial features;
2. made mistakes for some of the other ones.

justinmk commented:
> > I don't see any other enable() methods in :help lsp. Why is this the only one?
> 
> Because we (1) started with the simple, non-controversial features ... (2) made mistakes for some of the other ones.

To avoid more mistakes can we look at `vim.diagnostic.config`? Instead of adding lots of `foo_enable()` functions, it sounds like we need `vim.lsp.config` that follows the model of `vim.diagnostic.config`.

dchinmay2 committed:
include link to spec in news.txt

clason commented:
I believe the possibility of (easily) toggling diagnostics is something that's been missing from that API as well. I do agree (as I've written before) that it makes sense to look at this (toggling "code overlays") as a general issue.

mfussenegger commented:
> Does LSP actually require the enable/disable stuff to be configured client-side? Seems like this could/should be a setting that could be sent to the server, using a standard "configure" LSP request. Does LSP have something like that? Can the client-side "capabilities" map be used to enable/disable inlay hints?

Given that inlay hints are requested by the client it's also up to the client to control it. The server has no business controlling client behaviour

> When I suggested skipping the public API in https://github.com/neovim/neovim/pull/23736 , my thinking was: just enable it by default. It doesn't need to be configurable.

Inlay hints was always the first thing I turned off in intellij as it's very intrusive. We can discuss the default but it definitely needs an option to turn it off.  And I'd say not just the display but the requests as well.  Users who don't use it shouldn't need to spend processing power on it.

dchinmay2 committed:
unify `__buffers` and `hint_cache_by_buf`

dchinmay2 committed:
use callback bufnr in `on_lines` callback, bufstate: remove __index override

dchinmay2 committed:
fix update condition

dchinmay2 committed:
remove `or 0` from resolve bufnr calls

resolve bufnr checks for nil now

justinmk commented:
> An option could be to extend vim.lsp.start/start_client with flags for the different features. E.g. `start({ ..., formatting = false, semantic_tokens = false, inlay_hints = true, ... })`

That sounds like a good compromise since it's already the current pattern. That avoids numerous `foo_enable` functions and gives us time to decide if `vim.lsp.config` is needed/acceptable.

clason commented:
> That avoids numerous foo_enable functions and gives us time to decide if vim.lsp.config is needed/acceptable.

No. The `foo_enable` functions are about _runtime_ choices, not "all or nothing". These are two separate issues that we should not conflate.

justinmk commented:
> We have so far only ever considered options 1 and 2, because we got away with it. But this feature crosses a line where option 3 becomes important

With a `vim.start()` parameter it is possible to mimic a "runtime" choice by stopping/restarting the server. In any case that can be a followup, it doesn't need to block the minimal approach, which is what we should start with.

clason commented:
> With a vim.start() parameter it is possible to mimic a "runtime" choice by stopping/restarting the server. 

Which may require multiple seconds for starting up and re-indexing the project... 

I sincerely believe that "runtime choice" is the primary use case for inlay hints, different from all prior LSP features we added (and would like to point out that this already _is_ the followup).

Anyway, as long as we are clear on what the ask here is, I will hold my peace from now on.

justinmk commented:
> I sincerely believe that "runtime choice" is the primary use case for inlay hints,

I could be out of touch, but I didn't think that toggling inlay hints on/off at will was a common thing. But again (1) the door is open for that after we decide a long-term interface, (2) doesn't need to block the minimal approach.

p00f commented:
@justinmk I think vscode has a thing where it shows inlay hints as long as you hold certain keys

Not saying this should be the reference, just addressing
> didn't think that toggling inlay hints on/off at will was a common thing

gpanders commented:
> > I sincerely believe that "runtime choice" is the primary use case for inlay hints,
> 
> I could be out of touch, but I didn't think that toggling inlay hints on/off at will was a common thing. But again (1) the door is open for that after we decide a long-term interface, (2) doesn't need to block the minimal approach.

FWIW I didn’t think that either until I tried it. And my almost immediate thought was that I wanted to be able to flip them on when needed, but disable them otherwise. Personally, I find them far too noisy to leave enabled all the time.

justinmk commented:
@mfussenegger 

> A user could change the server capability, to trick the client into thinking it's not supported

But the client also declares its capabilities (`vim.lsp.start({capabilities={...}})`), don't see why a "trick" is needed.

> An option could be to extend vim.lsp.start/start_client with flags for the different features. E.g. `start({ ..., formatting = false, semantic_tokens = false, inlay_hints = true, ... })` 

Let's forget that.  [coc.nvim](https://github.com/neoclide/coc.nvim/commit/2bd87f66c7048e6c7c3a3d067be5c5e6e0036d33) for example, suggests there is a ridiculous amount of configuration people want to fiddle with. Can't jam that all into `vim.lsp.start()`, and see below...

> A `vim.lsp.config` could act as a skeleton, but it would need a way to distinguish between settings for _all_ clients or specific clients.

`vim.lsp.get_active_clients()[1]` shows me:

```
{
  config = {
    _on_attach = <function 3>,
    autostart = true,
    capabilities = { … },
    settings = {
      Lua = {
        diagnostics = {
          globals = { "vim" }
        },
        runtime = { … },
      }
    },
  },
  server_capabilities = { … },
  ...
}
```

Why isn't this the interface for setting client config?

    vim.lsp.get_client().config.settings.inlay_hints = true

I can imagine some cases where "current buf has multiple clients", then we might need: 

    vim.lsp.buf_clients.config.settings.inlay_hints = true

dchinmay2 committed:
move user-facing functions into vim.lsp.buf, unify enable/disable/toggle

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -1316,6 +1316,13 @@ incoming_calls()                                *vim.lsp.buf.incoming_calls()*
+
+    Parameters: ~
+      • {enable}  (boolean|nil) true/false to enable/disable, nil to toggle
+      • {bufnr}   (integer|nil) Buffer handle, or nil for current
```
Convention is:

- 0 = current buffer
- `nil` = all buffers (don't necessarily need to implement this here)

p00f replied:
done

clason replied:
Would be nice to do this in a follow-up, though, as now you can't just map to `vim.lsp.buf.inlay_hints()` (like you can with most other `vim.lsp.buf` functions).

dchinmay2 committed:
use 0 instead of nil for current buffer

dchinmay2 committed:
change docs to reflect nil -> 0

justinmk reviewed (approved):
LGTM (after CI is passing), will wait for other approvals, ping after a few days if not merged already.

Thanks for your persistence and patience!

dchinmay2 committed:
fix tests

mfussenegger reviewed (commented):
Approved this, but I think we should look into doing some API adjustments before this gets released.


Inlay hints have optional `textEdits`, and we could (or should?) allow users to apply them. That will likely need another function. With the current approach this would mean something like `vim.lsp.buf.inlay_hints_run()` (or `apply` or `apply_edits` or whatever). Which would just forward to `_inlay_hints.xy()`.

I'm thinking maybe we could go with a callable module approach similar to how `vim.iter` works. We'd have `vim.lsp.inlay_hints()` to toggle, enable or disable and could add additional functionality as needed in `vim.lsp.inlay_hints.xy()`.
This would remove the indirection/forwarding to a private module, and we could keep stuff properly encapsulated via lua `local`s and it's module system.


The same approach would also work for `codelens`. We could toggle/enable/disable via `vim.lsp.codelens()`, but we still have `vim.lsp.codelens.on_codelens` for the handler and `vim.lsp.codelens.run()` to execute a lens.
